src/main/java/org/ibp/api/java/impl/middleware/program/ProgramServiceImpl.java

```diff
@@ -113,6 +115,33 @@ public long countProgramsByFilter(final ProgramSearchRequest programSearchReques
 		return this.programService.filterPrograms(programSearchRequest, pageable);
 	}
 
+	@Override
+	public ProgramDTO getLastOpenedProject(final Integer userId) {
+		final Project project = this.workbenchDataManager.getLastOpenedProject(userId);
+		return project != null ? new ProgramDTO(project) : null;
+	}
+
+	@Override
+	public void saveOrUpdateProjectUserInfo(final Integer userId, final Long  projectId) {
+		final WorkbenchUser user = this.userService.getUserById(userId);
+		final Project project = this.workbenchDataManager.getProjectById(projectId);
+
+		final ProjectUserInfo projectUserInfo =
```

Comment:
This logic where the hibernate entity is built should better go into Middleware instead of BMSAPI